### PR TITLE
Added option to ignore specific metrics when checking max-series-per-metric limit.

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR adds new option `-ingester.ignore-series-limit-for-metric-names` to ignore specific metric names when checking series-per-metric limits.

This is useful under some very specific scenarios, when small number of metric names is used for ingesting series from non-Prometheus system. This is unlikely to be used in generic Cortex configurations.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -49,6 +49,7 @@
   * `cortex_ruler_write_requests_failed_total`
   * `cortex_ruler_queries_total`
   * `cortex_ruler_queries_failed_total`
+* [ENHANCEMENT] Ingester: Added option `-ingester.ignore-series-limit-for-metric-names` with comma-separated list of metric names that will be ignored in series-per-metric limit. #4302
```
Comment (pracucci) [inline CHANGELOG.md:52]: [nit] "series-per-metric" doesn't perfectly match a CLI flag or YAML config option, so I would just write it in plain english.

```suggestion
* [ENHANCEMENT] Ingester: Added option `-ingester.ignore-series-limit-for-metric-names` with comma-separated list of metric names that will be ignored in max series per metric limit. #4302
```

File: pkg/ingester/user_state.go [hunk 2]
```diff
@@ -362,9 +362,11 @@ type metricCounterShard struct {
 type metricCounter struct {
 	limiter *Limiter
 	shards  []metricCounterShard
+
+	ignoredMetricsForSeriesCount map[string]struct{}
```
Comment (pracucci) [inline pkg/ingester/user_state.go:366]: [nit] I think we can shorten the name. This struct only count series by metric name.
```suggestion
	ignoredMetrics map[string]struct{}
```

**Review by pracucci — APPROVED** [review]
LGTM!

**Review by stevesg — APPROVED** [review]
This configuration should only be used if all tenants are "trusted", correct? As anyone would have the ability to pass the limit by guessing the correct metric name. In a shared environment, it might make sense to be configurable per-tenant, but I see it wouldn't be worth the effort right now until there is a use case.

[non blocking nit] We could we document the option that it's not recommended for use with untrustworthy tenants.

**Comment by pstibrany** [comment]
> This configuration should only be used if all tenants are "trusted", correct? As anyone would have the ability to pass the limit by guessing the correct metric name. In a shared environment, it might make sense to be configurable per-tenant, but I see it wouldn't be worth the effort right now until there is a use case.
>
> [non blocking nit] We could we document the option that it's not recommended for use with untrustworthy tenants.

If tenant guesses the right metric name, they can avoid the per-metric series limit. This however isn't very useful... while such tenant would be able to create too many series for this metric, querying the metric back via PromQL would be difficult. And tenants have little incentive to rename all their metrics to use single metric name. Note that this option doesn't allow them to avoid tenant's series count limit.

My original design added this as overrideable (per-tenant) config option, however it complicated the design:
- I wanted to keep option as string, but return map for quick lookups
- Building map on first use in `*Limits` requires some concurrency control. However this (eg. sync.Mutex, or atomic pointer) cannot be stored in `*Limits` because we copy this structure for unmarshalling, and that breaks these primitives.
- After some fighting with it and trying few different designs, I chose simple solution of not using per-tenant option.
